doc/sphinxext/move_gallery_links.py

```diff
@@ -141,22 +145,31 @@ def _create_badge_link(link):
                 # These will be removed at the end if new links are successfully created
                 sg_note = soup.find("div", class_="sphx-glr-download-link-note")
                 sg_footer = soup.find("div", class_="sphx-glr-footer")
+                assert sg_note is not None, "Failed to find the top note"
+                assert sg_footer is not None, "Failed to find the gallery footer"
```

Comment:
I have the feeling that those assertions might still fail if we run some examples, but not all via the `EXAMPLES_PATTERN` env variable: https://scikit-learn.org/stable/developers/contributing.html#building-the-documentation

Can you please check?

Reply:
Ah didn't notice that because in `make.bat` we did not have this option. I will take it into account.

Reply:
I removed everything and built from scratch on my local machine with `html-noplot`, but it seems that I did not get those errors?

Then I looked at the `sphinx-gallery` source code: in the [`save_rst_example` function](https://github.com/sphinx-gallery/sphinx-gallery/blob/2111ee6a10d287c3f8e65dc989db384309399f52/sphinx_gallery/gen_rst.py#L1395) no matter whether the example is run or not the components we tweak will be added.

Is it possible that we are using different versions of `sphinx-gallery` (mine is `0.15.0`)? Or can you please provide more context on your side @ogrisel?

Reply:
Let me try again with a clean environment and recent versions of everything.

Reply:
So I created a new env with the latest sphinx-gallery (0.15.0) and ran `make html-noplot` and I can reproduce the `WARNING: Failed to tweak gallery links in...` warnings I initially reported in https://github.com/scikit-learn/scikit-learn/pull/28331#issuecomment-1959582859.

Also, I atried this branch and the warning go away. But it seems that I can no longer get the `EXAMPLES_PATTERN` thing to work. For instance I tried `EXAMPLES_PATTERN=".*mean.*" make html-noplot` and I did not get the plot generated in `plot_kmeans_stability_low_dim_dense.html`. I don't remember how it's supposed to work. But doing experiment is very slow. It seems that `make html-noplot` is much slower than it used to be. We would need to profile to find why it's the case.

Reply:
> So I created a new env with the latest sphinx-gallery (0.15.0) and ran `make html-noplot` and I can reproduce the `WARNING: Failed to tweak gallery links in...` warnings I initially reported in [#28331 (comment)](https://github.com/scikit-learn/scikit-learn/pull/28331#issuecomment-1959582859).

Thanks for experimenting! Then there might some mistakes made on my side. I will take another look into it some time.

> Also, I atried this branch and the warning go away. But it seems that I can no longer get the `EXAMPLES_PATTERN` thing to work. For instance I tried `EXAMPLES_PATTERN=".*mean.*" make html-noplot` and I did not get the plot generated in `plot_kmeans_stability_low_dim_dense.html`. I don't remember how it's supposed to work.

Hmm that's strange. I don't think it's likely caused by this PR because `tweak_gallery_links` is just a post transform that executes in the very last step while plot generation is determined very early even before page generation. I will try to test this as well.

> But doing experiment is very slow. It seems that `make html-noplot` is much slower than it used to be. We would need to profile to find why it's the case. Note: I have not yet tried on `main`, maybe the slowness is already there.

I also noticed the significant slowdown (as long as using `pydata-sphinx-theme`): not 100% sure but it may be caused by https://github.com/pydata/pydata-sphinx-theme/issues/1643.